README.md
# The Magic Elevator

## Team Members
 - Max Funck
 - Shawn Polson
 
## Video Demo
[![Magic Elevator Demo Video](https://user-images.githubusercontent.com/14846863/57108134-18c41d00-6cef-11e9-9650-c58e91006778.jpg)](https://youtu.be/wAoRdG36DOk)
https://youtu.be/wAoRdG36DOk


## The Idea
Elevators are both a blessing and a curse. They enable heavy equipment to be transported across levels, and they are often the only means by which people with mobility impairments can move up and down in a building. However, waiting for them in an awkward silence is anything but exciting. The Magic Elevator solves this problem: riders become witches and wizards, summoning their elevator through magic spells. Only a correctly executed spell calls the elevator!

This is a minimum viable product (MVP) prototype of a system that accomplishes this idea. The prototype was built by Maximilian Funck and Shawn Polson as part of their Machine Learning for Human-Computer Interaction course (CSCI 5880) at CU Boulder. Witches/wizards wield magic wands that are plastic toy Harry Potter wands whose insides were gutted and replaced with a micro:bit that senses X, Y, and Z acceleration. They perform wand motions with voice incantations that together call the elevator either up or down. The witches/wizards come up with and train the motions and incantations themselves through the Magic Elevator's user interface (UI). When spells are cast correctly, a button-pushing mechanism built almost entirely out of LEGO parts (and a Raspberry Pi 3), which is mounted over the elevator buttons, summons the elevator in the appropriate direction.

Witches/wizards are free to train a new wand motion or set a new incantation for their spells at any time; Professor Flitwick walks them through it in the UI. When two users are present, they must work together to cast agreeing spells to call the elevator in one direction. It's a game; only sufficiently skilled witches/wizards may enter the Magic Elevator! When only one is present, she or he can just call the elevator in single-player mode.

## Mission Statement
Our mission is to provide a Harry Potter–like elevator experience with machine learning technology. The interaction with the input device, a wand, results in a magically activated elevator terminal pressing the buttons to call the elevator. The user experience is designed to be seamless and intuitive, just like magic.

## User Experience
![UserStoryGithub](https://user-images.githubusercontent.com/46902147/56765798-20227e00-6765-11e9-979d-04eb72f43116.png)  

## Building the Project
This repo contains all the software needed to recreate the Magic Elevator. The hardware, wands and the mechanical button pusher, were custom-built and recreating them isn't trivial. But here we include details for building all aspects of our system. 

### Building the Software
![UserInterface](https://user-images.githubusercontent.com/14846863/56779108-adca9180-6796-11e9-9332-b05396f1e6eb.png)

All the software is shown in the picture above. Here's how to use/build each component.
#### Main UI
The program labeled "Main UI" in the picture above is [`Flitwick_Spell_Training/Flitwick_Spell_Training.pde`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/Flitwick_Spell_Training/Flitwick_Spell_Training.pde) in this repo. It's a self-contained Processing program that will run once you have installed the libraries it imports. It relies on three system fonts that we put in [`Flitwick_Spell_Training/data/`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/Flitwick_Spell_Training/data): `4PrivetDrive-Regular.otf`, `HARRP___.TTF`, and `LumosLatino.ttf`.

#### Wand data
The script labeled "Wand data" is acceleration output from the wand micro:bits. Three code files are needed to stream this data: [`wand/mbit/microbit-sendXYZ*.hex`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/wand/mbit), [`wand/mbit/microbit-receiveXYZ*.hex`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/wand/mbit), and [`wand/py/wand_to_osc.py`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/wand/py/wand_to_osc.py). The data streams from the micro:bit in a wand to a micro:bit plugged into the computer, then to `wand_to_osc.py` which prints and sends the data. Building the wand itself is covered in the **Building the Hardware** section. To make the software work, plug the wand micro:bit into your computer and drag either `microbit-sendXYZ.hex` or `microbit-sendXYZ2.hex` into the MICROBIT directory that appears. There are two versions of this send code, and of the receive code for that matter, to support using two wands at once; the only difference between the files is that they use different radio channels. After you drag the send code into the wand micro:bit, plug in a second microbit and drag either `microbit-receiveXYZ.hex` or `microbit-receiveXYZ2.hex`, depending on which send version you use in the wand, into the MICROBIT directory that appears. You are now streaming acceleration data from the wand to your computer. The last step is to run `wand_to_osc.py` after changing line 10 match the serial port your computer generates for the receiving micro:bit. You will then see the data streaming like in the picture.

#### Classify wand motions
The program labeled "Classify wand motions" is [Wekinator](http://www.wekinator.org/), free, open-source software for real-time, interactive machine learning. It uses a dynamic time warping model to classify wand motions from the wand data. You could learn about Wekinator from the documentation on their website and configure your own model, but for convenience, we've uploaded the model we used in [`WandToMotorModel/`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/WandToMotorModel). Download Wekinator, open it, click `File -> Open project...` and select the [`WandToMotorModel.wekproj`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/WandToMotorModel/WandToMotorModel.wekproj) file. This model will not have any training examples for wand motions, so you will want to train your own motions for calling the elevator up and down by clicking the appropriate buttons in the main UI. `wand_to_osc.py` sends the wand's acceleration data to Wekinator via OSC messages sent to localhost on port 8999. Make sure Wekinator is listening on that port (our uploaded model will be). Wekinator sends its classifications to the Raspberry Pi controlling the button pusher also via OSC messages, sent to *an IP address* at port 12000. You will have to change the IP address used for Wekinator's output to match the IP of your Raspberry Pi.

#### Transcribe incantations
This script is [`Flitwick_Spell_Training/incantations.py`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/Flitwick_Spell_Training/incantations.py) in this repo. It uses Google's speech-to-text API, via the Python library `speech_recognition`, to transcribe your incantations. Install the libraries it imports, then run it and watch it recognize phrases you speak. There are two caveats to using this script: 1) It must remain inside the `Flitwick_Spell_Training` directory because it reads incantations you set in the main UI from two files: [`data/up.txt`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/Flitwick_Spell_Training/data/up.txt) and [`data/down.txt`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/Flitwick_Spell_Training/data/down.txt). 2) You must change the IP address on line 5 of this script to match the IP address of your Raspberry Pi, so that the OSC messages it sends actually get delivered. 

#### Change incantations
This script is [`/Flitwick_Spell_Training/change_incantations_server.py`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/Flitwick_Spell_Training/change_incantations_server.py) in this repo. It must be running before you press either "Change Incantation" button in the main UI. This is because the first phrase it hears you speak after clicking one of those buttons gets written to either `up.txt` or `down.txt`—this is how we support changing incantations during runtime.

#### Receive spells
This program is [`pi/callelevator/callelevator.pde`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/pi/callelevator/callelevator.pde) in this repo. OSC messages from wand motions and voice incantations are received by this program. It must be running on your Raspberry Pi for the button pusher to work. 

#### LEGO motor control
This script is either [`pi/motor_osc_server.py`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/pi) or [`pi/motor_osc_server_2.py`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/tree/master/pi) in this repo. Running the former will put the Magic Elevator system in single-player mode, while the latter will put it in two-player mode. The difference is that in two-player mode, only two spells cast properly in agreement will call the elevator. When either script is running, it receives OSC messages from `callelevator.pde` that tell it to move the LEGO motor to press either the up or down elevator button.

### Building the Hardware

The hardware of the Magic Elevator consists of an input device—a magic wand—and an output device—the mechanical button pusher. The magic wand uses a [toy wand](https://www.amazon.com/Harry-Potter-Potters-Wizard-Training/dp/B01N29ENXW/ref=sr_1_3?keywords=harry+potter+wand&qid=1556837733&s=gateway&sr=8-3) as a case. The interior is taken out and replaced by a [micro:bit](https://www.amazon.com/BBC-BBC2546862-Micro-bit-go/dp/B01G8X7VM2/ref=sr_1_3?keywords=microbit&qid=1556837891&s=gateway&sr=8-3), and the battery of the micro:bit is stored in the handle of the wand. The wand can be seen in the picture below. 

<img src="https://user-images.githubusercontent.com/14846863/57108237-5759d780-6cef-11e9-82b4-d9cfac3aa789.PNG" width="400" title="Wand"></img>

The mechanical button pusher is responsible for physically pressing the elevator buttons. All hardware is mounted to a laser-cut acrylic sheet, shown in the picture below and available in this repo under [`ElevatorAcrylicSheet.sldprt`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/ElevatorAcrylicSheet.sldprt). In the final assembly, [3M Dual Lock tape](https://www.amazon.com/3M-Dual-Lock-Reclosable-Fastener/dp/B007OXK330/ref=sr_1_1_sspa?keywords=3m+dual+lock&qid=1556838467&s=gateway&sr=8-1-spons&psc=1) attaches the mechanism to the elevator panel.
![Bildschirmfoto 2019-05-02 um 4 51 45 PM](https://user-images.githubusercontent.com/46902147/57112208-4368a280-6cfc-11e9-9dce-231ff657809e.png)  

A LEGO assembly is built on the acrylic sheet. Since many different variations of elevator panels exist, we will forgoe a detailed construction manual. However, two critical LEGO components have proven to be suitable for the assembly. First, the LEGO MINDSTORMS Interactive Motor (9842), and second, two LEGO TECHNIC mini linear actuators (92693). Working with gears and belts, the clockwise and counterclockwise rotational movement of the motor can be transferred to extend the two linear actuators in opposing directions. These two LEGO components are shown in the picture below.

![LEGO](https://user-images.githubusercontent.com/46902147/57112560-ce966800-6cfd-11e9-9994-8441402ade01.png)

In addition, the acrylic sheet holds the mentioned Raspberry Pi 3 and BrickPi3 shield. A 12 Volt battery pack can be included to drive the Raspberry Pi, BrickPi, and LEGO motor. However, this load does drain the battery pack quickly. A 12 Volt wall plug is the more stable option. The final assembly is shown in the picture below.

![HardwareDetail](https://user-images.githubusercontent.com/46902147/56911665-a8a56500-6a6a-11e9-80de-c8a03416ab04.png)
 
## Final Report
[`final_report.md`](https://github.com/CUBoulder-2019Sp-IML4HCI/Final-Magic-Elevator-Funck-Polson/blob/master/final_report.md) in this repo.


changelog.md
## [0.0.1] – 2019-03-12
### Added
 - Max reset the software in the Raspberry Pi, wired the Pi back into our 3D-printed housing, and reconnected it to our solenoids.
 - Max got Zerotier connectivity working between the Raspberry Pi and both his and Shawn's laptops.
 - Both Max and Shawn can now remotely connect to and control the Raspberry Pi from their laptops.
 - Max confirmed that simple classification OSC messages from Wekinator can be recieved by the Raspberry Pi.
 
## [0.0.2] – 2019-03-15
### Added
 - Shawn rewrote the Processing code on the Raspberry Pi to accept DTW wand gestures from Wekinator, as opposed to simple classification OSC messages.
 - Shawn created, trained, and saved a Wekinator project that reliably recognizes "up" and "down" spells from a wand. 
 - Shawn confirmed that wand spells can in fact cause our solenoids to fire.
 
## [0.0.3] – 2019-03-19
### Added
 - Both Max and Shawn tested our solenoid system on the elevator in Atlas, and all the components work with one another, but the solenoids just are not powerful enough to actually press the elevator buttons!
 - Shawn uploaded all relevant code files to GitHub and created this changelog file.
 - Both Max and Shawn recorded this progress update video: https://youtu.be/D2E1a5-l_-8

## [0.0.4] – 2019-04-03
### Added
 - Shawn added a voice recognition feature to the software. From now on, the elevator will only be called if both the gesture and the corresponding outspoken spell are correctly recognized. 
 - Max researched alternatives to the previously tested solenoids. Other customers of that solenoid type also reported that the solenoid only produces about half the force than stated in the datasheet. Alternatives are A: stronger solenoids B: electric linear motor or C: a pneumatic system. The specifications of each option can be found in the proposal document. All options drive costs and complexity of the system while reducing the mobility.
 - Both Shawn and Max will discuss hacking options with facility management on 04/04/2019.
 - Shawn recorded this progress update video: https://youtu.be/4teWFxSjB8g

## [0.0.5] – 2019-04-09
### Added
 - Shawn altered the voice-to-text script to send OSC messages when key spell phrases are in the returned text, as opposed to the returned text matching the phrases exactly
 - Max built a linear actuator out of LEGO parts that is finally powerful enough to press elevator buttons
 - Both Shawn and Max recorded this progress update video: https://youtu.be/d8HVFMx6nRo
 
## [0.0.6] – 2019-04-16
### Added
 - Shawn added a Python script that controls Wekinator with voice commands. Specifically, it can control starting/stopping running and recording DTW examples. He also wrote a different script that walks the user through training an "up" gesture without the need for voice commands.
 - Max designed and laser cut acrylic sheets for our hardware. Further iterations of the hardware to avoid "pushing itself off the wall" have been made. Gears have been modified to activate both buttons with one motor.
 - Both Shawn and Max received the BrickPi3 and set it up.
 - Shawn got the previous software working with the new LEGO hardware (including Zerotier, VNC connection and LEGO motor activation).
 - Both Shawn and Max recorded this progress update video: https://youtu.be/z2Na-lmQrTQ
 
## [0.0.7] – 2019-04-25
### Added
 - Shawn finished the wand motion training/incantation changing user interface (making it look and feel very much like Harry Potter)
 - Max wrote the first draft of the final project report
 - Both Max and Shawn revised the final project report
 
## [0.1.0] – 2019-04-27
### Added
 - Both Max and Shawn finished the final project report (`final_report.md` in this repo)
 - Both Max and Shawn finished the final project presentation (`final_presentation.txt` in this repo)
 - Both Max and Shawn finished the final project video: https://youtu.be/wAoRdG36DOk


test_train_up_gesture.py
import time
from pythonosc import udp_client


client = udp_client.SimpleUDPClient("localhost", 8999)

   
print("Okay! Let's train a new gesture for UP.")
time.sleep(2)
print("I will erase the current gesture in one moment...")
time.sleep(3)
client.send_message("/wekinator/control/stopRunning", 0 ) # stop running
client.send_message("/wekinator/control/deleteExamplesForOutput", 1 ) # delete all UP gestures
print("Erased! Now let's train that new gesture.\n")
time.sleep(2)

print("Get your wand ready.")
time.sleep(2)
print("I'll count down from 5, and once I say GO! I'll have you perform the gesture 10 times.")
print("You'll have 2 seconds each time.")
time.sleep(5)
print("Ready?")
time.sleep(2)

print("5...")
time.sleep(1)
print("4...")
time.sleep(1)
print("3...")
time.sleep(1)
print("2...")
time.sleep(1)
print("1...")
time.sleep(1)
print("\nGO!")

for i in range(1, 11):
    print("START GESTURE (" + str(i) + ")")

    client.send_message("/wekinator/control/startDtwRecording", 1 )  # start recording UP
    time.sleep(2)  # record for exactly 2 seconds

    print("STOP GESTURE")
    client.send_message("/wekinator/control/stopDtwRecording", 0 ) # stop recording UP
    time.sleep(1.5)

print("Great spell casting! Now try out your new gesture.")

client.send_message("/wekinator/control/startRunning", 0 )  # start running the model again




            


pi/motor_osc_server.py
#!/usr/bin/env python
#
# 
# This code runs the NXT motor to move it in a postion that presses either the up or down elevator button, depending on the OSC message received. 
# Hardware: Connect the NXT motor to the BrickPi3 motor port A. Make sure that the BrickPi3 is running on a 9v power supply.
#
# Results:  When you run this program, send OSC messages to it and the motor will move correspondingly.

from __future__ import print_function # use python 3 syntax but make it compatible with python 2
from __future__ import division       #   

from pythonosc import dispatcher
from pythonosc import osc_server
import argparse
import sys      #                        ''

import time     # import the time library for the sleep function
import brickpi3 # import the BrickPi3 drivers


class MotorController:
    
    def __init__(self, bp):
        self.BP = bp
        # Positions assume motor is at 0 when code starts
        self.press_up = -3050    
        self.press_down = 2700
        print("Initialized.")


    def call_up(self, signal_name, data):
        #print("Called up!")
        #print("signal_name:")
        #print(signal_name)
        #print("data:")
        #print(data)
        target = self.press_up
        BP.set_motor_position(BP.PORT_A, target)    # set motor A's target position to the current position of motor D
        print("Pressed UP: Motor A target: %6d  Motor A position: %6d" % (target, BP.get_motor_encoder(BP.PORT_A)))
        time.sleep(10.0)  # delay for 2.0 seconds
        BP.set_motor_position(BP.PORT_A, 0)  # reset motor position    
            
    def call_down(self, signal_name, data):
        #print("Called down!")
        #print("signal_name:")
        #print(signal_name)
        #print("data:")
        #print(data)
        target = self.press_down
        BP.set_motor_position(BP.PORT_A, target)    # set motor A's target position to the current position of motor D
        print("Pressed DOWN: Motor A target: %6d  Motor A position: %6d" % (target, BP.get_motor_encoder(BP.PORT_A)))
        time.sleep(10.0)  # delay for 2.0 seconds
        BP.set_motor_position(BP.PORT_A, 0)  # reset motor position

         
    
    
if __name__ == "__main__":
    
    BP = brickpi3.BrickPi3() # Create an instance of the BrickPi3 class. BP will be the BrickPi3 object.

    try:
        BP.offset_motor_encoder(BP.PORT_A, BP.get_motor_encoder(BP.PORT_A)) # reset encoder A
    except IOError as error:
        print(error)

    BP.set_motor_limits(BP.PORT_A, 99, 400)          # optionally set a power limit (in percent) and a speed limit (in Degrees Per Second)
    
    ctrl = MotorController(BP)

    ip = '127.0.0.1'  # pi ip: 192.168.195.28
    port = '12001'
    
    # Parse arguments for host ip and port.
    parser = argparse.ArgumentParser()
    parser.add_argument("--ip",
        default=ip, help="The ip to listen on")
    parser.add_argument("--port",
        type=int, default=port, help="The port to listen on")
    args = parser.parse_args()
    
    # Create dispatcher to handle incoming OSC messages.
    dispatcher = dispatcher.Dispatcher()
    dispatcher.map("/elevator/up", ctrl.call_up)
    dispatcher.map("/elevator/down", ctrl.call_down)
    
    server = osc_server.ThreadingOSCUDPServer(
            (args.ip, args.port), dispatcher)
        
    print("Serving on {}".format(server.server_address))
    server.serve_forever()




pi/motor_osc_sever_2.py
#!/usr/bin/env python
#
# 
# This code runs the NXT motor to move it in a postion that presses either the up or down elevator button, depending on the OSC message received. 
# Hardware: Connect the NXT motor to the BrickPi3 motor port A. Make sure that the BrickPi3 is running on a 9v power supply.
#
# Results:  When you run this program, send OSC messages to it and the motor will move correspondingly.

from __future__ import print_function # use python 3 syntax but make it compatible with python 2
from __future__ import division       #   

from pythonosc import dispatcher
from pythonosc import osc_server
import argparse
import sys      #                        ''

import time     # import the time library for the sleep function
import brickpi3 # import the BrickPi3 drivers


class MotorController:
    
    def __init__(self, bp):
        self.BP = bp
        # Positions assume motor is at 0 when code starts
        self.press_up = -3050    
        self.press_down = 2700
        self.num_ups = 0
        self.num_downs = 0
        print("Initialized LEGO motor for two people.")


    def call_up(self, signal_name, data):
        self.num_ups = self.num_ups + 1
        self.num_downs = 0

        if (self.should_call_elevator('up') == True):
            target = self.press_up
            BP.set_motor_position(BP.PORT_A, target)    # set motor A's target position to the current position of motor D
            print("Pressed UP: Motor A target: %6d  Motor A position: %6d" % (target, BP.get_motor_encoder(BP.PORT_A)))
            time.sleep(10.0)  # delay for 10.0 seconds
            BP.set_motor_position(BP.PORT_A, 0)  # reset motor position    
            
    def call_down(self, signal_name, data):
        self.num_downs = self.num_downs + 1
        self.num_ups = 0

        if (self.should_call_elevator('down') == True):
            target = self.press_down
            BP.set_motor_position(BP.PORT_A, target)    # set motor A's target position to the current position of motor D
            print("Pressed DOWN: Motor A target: %6d  Motor A position: %6d" % (target, BP.get_motor_encoder(BP.PORT_A)))
            time.sleep(10.0)  # delay for 10.0 seconds
            BP.set_motor_position(BP.PORT_A, 0)  # reset motor position

    def should_call_elevator(self, direction):
        if (direction == 'up'):
            if (self.num_ups > 1):
                print('Received 2nd UP spell.')
                self.num_ups = 0
                return True
            else:
                print('Received 1st UP spell.')
                return False

        elif (direction == 'down'):
            if (self.num_downs > 1):
                print('Received 2nd DOWN spell.')
                self.num_downs = 0
                return True
            else:
                print('Received 1st DOWN spell.')
                return False

        else:
            return False

         
        
if __name__ == "__main__":
    
    BP = brickpi3.BrickPi3() # Create an instance of the BrickPi3 class. BP will be the BrickPi3 object.

    try:
        BP.offset_motor_encoder(BP.PORT_A, BP.get_motor_encoder(BP.PORT_A)) # reset encoder A
    except IOError as error:
        print(error)

    BP.set_motor_limits(BP.PORT_A, 99, 400)          # optionally set a power limit (in percent) and a speed limit (in Degrees Per Second)
    
    ctrl = MotorController(BP)

    ip = '127.0.0.1'  # pi ip: 192.168.195.28
    port = '12001'
    
    # Parse arguments for host ip and port.
    parser = argparse.ArgumentParser()
    parser.add_argument("--ip",
        default=ip, help="The ip to listen on")
    parser.add_argument("--port",
        type=int, default=port, help="The port to listen on")
    args = parser.parse_args()
    
    # Create dispatcher to handle incoming OSC messages.
    dispatcher = dispatcher.Dispatcher()
    dispatcher.map("/elevator/up", ctrl.call_up)
    dispatcher.map("/elevator/down", ctrl.call_down)
    
    server = osc_server.ThreadingOSCUDPServer(
            (args.ip, args.port), dispatcher)
        
    print("Serving on {}".format(server.server_address))
    server.serve_forever()



Flitwick_Spell_Training/incantations.py
import speech_recognition as sr
from pythonosc import udp_client

r = sr.Recognizer()
ip = '192.168.195.28'  # zerotier IP of raspberry pi
client = udp_client.SimpleUDPClient(ip, 12000)

while(True):
    with open('data/up.txt') as uf:
        with open('data/down.txt') as df:
            with sr.Microphone() as source:
                print('Speak Anything: ')
                audio = r.listen(source)

                try:
                    up_incantation = uf.read().strip()
                    down_incantation = df.read().strip()
                    text = r.recognize_google(audio)

                    if (up_incantation in text):
                        print('UP MATCH')
                        client.send_message('/up_voice', 1 )
                    elif (down_incantation in text):
                        print('DOWN MATCH')
                        client.send_message('/down_voice', 2 )
                    
                    print('You said:  {}'.format(text))
                    print('')
                except Exception as e:
                    print('Sorry, could not understand your voice.')
                

Flitwick_Spell_Training/change_incantations_server.py
import speech_recognition as sr
from pythonosc import udp_client
from pythonosc import dispatcher
from pythonosc import osc_server
import argparse
import sys                      


class IncantationChanger:
    
    def __init__(self):
        print("Initialized incantation changer.")


    def changeUp(self, unused_data):
        with open('data/up.txt', 'w') as up_file:
            incantation = ''

            # get new incantation from user's voice
            r = sr.Recognizer()
            while(True):
                with sr.Microphone() as source:
                    print('Speak a new UP incantation: ')
                    audio = r.listen(source)

                    try:
                        text = r.recognize_google(audio)
                        incantation = text
                        
                        print('Your new UP incantation is:  {}'.format(incantation))
                        print('')
                        break
                    except Exception as e:
                        print('...')


            up_file.write(incantation)
            up_file.close()

    def changeDown(self, unused_data):
        with open('data/down.txt', 'w') as down_file:
            incantation = ''

            # get new incantation from user's voice
            r = sr.Recognizer()
            while(True):
                with sr.Microphone() as source:
                    print('Speak a new DOWN incantation: ')
                    audio = r.listen(source)

                    try:
                        text = r.recognize_google(audio)
                        incantation = text
                        
                        print('Your new DOWN incantation is:  {}'.format(incantation))
                        print('')
                        break
                    except Exception as e:
                        print('...')

            down_file.write(incantation)
            down_file.close()
        
  
    
if __name__ == "__main__":

    changer = IncantationChanger() 

    ip = '127.0.0.1'  
    port = '12002'
    
    # Parse arguments for host ip and port.
    parser = argparse.ArgumentParser()
    parser.add_argument("--ip",
        default=ip, help="The ip to listen on")
    parser.add_argument("--port",
        type=int, default=port, help="The port to listen on")
    args = parser.parse_args()
    
    # Create dispatcher to handle incoming OSC messages.
    dispatcher = dispatcher.Dispatcher()
    dispatcher.map("/incantation/up", changer.changeUp)
    dispatcher.map("/incantation/down", changer.changeDown)
    
    server = osc_server.ThreadingOSCUDPServer(
            (args.ip, args.port), dispatcher)
        
    print("Serving on {}".format(server.server_address))
    server.serve_forever()


voice_control_wekinator.py
import speech_recognition as sr
import time
from pythonosc import udp_client

r = sr.Recognizer()
client = udp_client.SimpleUDPClient("localhost", 8999)

while(True):
    with sr.Microphone() as source:
        print('Speak Anything: ')
        audio = r.listen(source)

        try:
            text = r.recognize_google(audio)

            # voice commads to start Processing code that trains new gestures
            if ('begin training up' in text):
                client = udp_client.SimpleUDPClient("localhost", 12000)
                client.send_message("/startTrainingUp", 0 )
                client = udp_client.SimpleUDPClient("localhost", 8999) 
            elif ('begin training down' in text):
                client = udp_client.SimpleUDPClient("localhost", 12000)
                client.send_message("/startTrainingDown", 0 )
                client = udp_client.SimpleUDPClient("localhost", 8999) 

            # assortment of voice commands to control Wekinator directly
            elif ('start running' in text ):
                client.send_message("/wekinator/control/startRunning", 0 )
            elif ('stop running' in text):
                client.send_message("/wekinator/control/stopRunning", 0 )
            elif ('start recording up' in text):
                client.send_message("/wekinator/control/startDtwRecording", 1 )
                time.sleep(2)  # record for exactly 2 seconds
                client.send_message("/wekinator/control/stopDtwRecording", 0 )
            elif ('start recording down' in text):
                client.send_message("/wekinator/control/startDtwRecording", 2 )
                time.sleep(2)  # record for exactly 2 seconds
                client.send_message("/wekinator/control/stopDtwRecording", 0 )
            elif ('start recording default' in text):
                client.send_message("/wekinator/control/startDtwRecording", 3)
                time.sleep(2)  # record for exactly 2 seconds
                client.send_message("/wekinator/control/stopDtwRecording", 0 )
            # elif ('stop recording' in text):
            #     client.send_message("/wekinator/control/stopDtwRecording", 0 )
            # TODO: "deleteExamplesForOutput" (with 1, 2, or 3) to support deleting bad DTW recordings?
            
            print('You said:  {}'.format(text))
        except Exception as e:
            print(e)
            print('Sorry, did not hear/recognize your voice.')
            
